pkg/evetestkit/utils.go

```diff
@@ -728,3 +731,121 @@ func InitilizeTest(projectName string, options ...TestOption) (*EveNode, error)
 
 	return rnode, nil
 }
+
+func NewTestContextFromConfig(cfg *openevec.EdenSetupArgs) (*testcontext.TestContext, error) {
+	var (
+		err       error
+		sdnClient *edensdn.SdnClient
+		withSdn   bool
+	)
+
+	devModel := cfg.Eve.DevModel
+	eveRemote := cfg.Eve.Remote
+	withSdn = !cfg.Sdn.Disable &&
+		devModel == defaults.DefaultQemuModel &&
+		!eveRemote
+	if withSdn {
+		sdnSSHPort := cfg.Sdn.SSHPort
+		sdnMgmtPort := cfg.Sdn.MgmtPort
+		sdnSourceDir := filepath.Join(cfg.Eden.Root, strings.TrimSpace(cfg.Sdn.SourceDir))
+		sdnSSHKeyPath := filepath.Join(sdnSourceDir, "vm/cert/ssh/id_rsa")
+		sdnClient = &edensdn.SdnClient{
+			SSHPort:    uint16(sdnSSHPort),
+			SSHKeyPath: sdnSSHKeyPath,
+			MgmtPort:   uint16(sdnMgmtPort),
+		}
+	}
+
+	vars, err := openevec.InitVarsFromConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	ctx := &controller.CloudCtx{Controller: &adam.Ctx{}}
+	ctx.SetVars(vars)
+	if err := ctx.InitWithVars(vars); err != nil {
+		return nil, err
+	}
+	ctx.GetAllNodes()
+	tstCtx := &testcontext.TestContext{
+		Cloud:     ctx,
+		Tests:     map[*device.Ctx]*testing.T{},
+		SdnClient: sdnClient,
+		WithSdn:   withSdn,
+	}
+	tstCtx.ProcBus = testcontext.InitBus(tstCtx)
+	return tstCtx, nil
+
+}
+
+func InitilizeTestFromConfig(projectName string, cfg *openevec.EdenSetupArgs, options ...TestOption) (*EveNode, error) {
+	var edgenode *device.Ctx
+	tc, err := NewTestContextFromConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Registering our own project namespace with controller for easy cleanup
+	tc.InitProject(fmt.Sprintf("%s_%s", projectName, time.Now()))
+
+	// Create representation of EVE instances (based on the names
+	// or UUIDs that were passed in) in the context. This is the first place
+	// where we're using zcli-like API:
+	for _, node := range tc.GetNodeDescriptions() {
```

Comment:
I'm a little confused by this function: why is it returning a slice, when Eden supports only one node?

Reply:
well, there was support for multinodes, the whole point of contexts was about multi-node support, if I remember it correctly

Reply:
but tc is actually next thing to refactor IMO, it's what's been used under the hood in escript, but now with simplified context of running (just one golang program) and the whole toolkit of golang, definitions, guidlines, books, etc etc. 

However, in this particular case I don't think it's a problem. We define interface, which we ideally want to have (i.e. that nice little black box which creates N edge nodes, sets up everything, makes really nice coffee, etc.) then we deal with implementation based on our sense of urgency of the features.

Granted, it's hard to implement ideal interface, that's why it's iterative process, but it's usually harder to change interface than an implementation which follows the same interface (or part of it)

Reply:
but even our openEVEC supports only one EVE config...
would be good to clean it up in the future - either full multi-node support or just single node

Reply:
@uncleDecart is right, the context is about an instance of EVE-OS which may be connected to the same Adam. Another context will use its own set of certificates to not conflict on Adam side. And we can switch between contexts from CLI. Nice to keep the ability.